### ISSUE TITLE
autoStartup(boolean) not functioning properly on Amqp.inboundAdapter

### ISSUE THREAD

**Issue by TabberGit** [spring-projects/spring-integration-java-dsl#80]
Running a very simple test application using **spring-integration-java-dsl-1.1.2.RELEASE**.

    return IntegrationFlows.from(
        Amqp.inboundAdapter(connectionFactory, "test.queue")
            .id("testInbound")
            .autoStartup(false))
        .<byte[]> handle((p, h) -> processData(p, h))
        .get();

having `autoStartup(false)` on the inboundAdapter does not prevent it from starting immediately as expected.

My team is trying to develop an app that will use the control bus to start/stop the endpoint (which, by the way works perfectly with things defined through DSL.)  The last bit of the puzzle is having the endpoint not begin processing as soon as the application starts.

**Comment by artembilan** [comment]
Hello, @TabberGit !

Thank you for coming here for consultation/help :smile: 

Well, I've just tested the case locally and it demonstrates good (bad ?) results:
```
@Bean
public IntegrationFlow amqpInboundFlow(ConnectionFactory rabbitConnectionFactory) {
	return IntegrationFlows.from((MessageProducers p) ->
			p.amqp(rabbitConnectionFactory, fooQueue())
					.autoStartup(false))
			.transform(String.class, String::toUpperCase)
			.channel(Amqp.pollableChannel(rabbitConnectionFactory)
					.queueName("amqpReplyChannel")
					.channelTransacted(true))
			.get();
}
```
See `AmqpTests` in the project code.

Since you say that it is started anyway in your application, let's try to figure out WHO (control bus ?) does that for you exactly in your application.

Please, configure your logger for this `org.springframework=DEBUG` and attach here the logs, when your `testInbound` is started. For the `autoStartup(true)` it looks like:
```
2016-05-09 14:42:28,242 DefaultLifecycleProcessor [main] : Starting beans in phase 1073741823
2016-05-09 14:42:28,242 DefaultLifecycleProcessor [main] : Starting bean 'myAmqpInbound' of type [class org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter]
2016-05-09 14:42:28,242 SimpleMessageListenerContainer [main] : No global properties bean
```
Although would be better to show logs as much as possible.

**Comment by TabberGit** [comment]
Thank you so much for your prompt response!

Interesting stuff here...  If i have `autoStartup(true)`, I see:

    2016-05-09 14:39:12.585 [main] INFO  org.springframework.context.support.DefaultLifecycleProcessor - Starting beans in phase 1073741823
    2016-05-09 14:39:12.585 [main] DEBUG org.springframework.context.support.DefaultLifecycleProcessor - Starting bean 'testInbound' of type [class org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter]
    2016-05-09 14:39:12.585 [main] DEBUG o.s.amqp.rabbit.listener.SimpleMessageListenerContainer - No global properties bean

and, if I have `autoStartup(false)`, I have:

    2016-05-09 14:42:10.000 [main] INFO  org.springframework.context.support.DefaultLifecycleProcessor - Starting beans in phase 1073741823
    2016-05-09 14:42:10.000 [main] DEBUG org.springframework.context.support.DefaultLifecycleProcessor - Starting bean 'testInbound' of type [class org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter]
    2016-05-09 14:42:10.001 [main] DEBUG o.s.amqp.rabbit.listener.SimpleMessageListenerContainer - No global properties bean

**Comment by TabberGit** [comment]
Here is everything that's said about *testInbound* (The name of my `AMQP.inboundAdapter`) in my logs on a `autoStartup(false)`:

    Line 14019: 2016-05-09 14:42:03.038 [main] DEBUG o.s.beans.factory.support.DefaultListableBeanFactory - Invoking afterPropertiesSet() on bean with name 'testInbound'
    Line 14830: 2016-05-09 14:42:05.270 [main] DEBUG o.springframework.web.servlet.handler.BeanNameUrlHandlerMapping - Rejected bean name 'testInbound': no URL paths identified
    Line 18476: 2016-05-09 14:42:07.412 [main] DEBUG o.s.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'testInbound'
    Line 18649: 2016-05-09 14:42:07.613 [main] DEBUG o.s.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'testInbound'
    Line 18663: 2016-05-09 14:42:07.630 [main] DEBUG o.s.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'testInbound'
    Line 18664: 2016-05-09 14:42:07.630 [main] INFO  org.springframework.integration.monitor.IntegrationMBeanExporter - Located managed bean 'org.springframework.integration:type=ManagedEndpoint,name=testInbound,bean=endpoint': registering with JMX server as MBean [org.springframework.integration:type=ManagedEndpoint,name=testInbound,bean=endpoint]
    Line 18664: 2016-05-09 14:42:07.630 [main] INFO  org.springframework.integration.monitor.IntegrationMBeanExporter - Located managed bean 'org.springframework.integration:type=ManagedEndpoint,name=testInbound,bean=endpoint': registering with JMX server as MBean [org.springframework.integration:type=ManagedEndpoint,name=testInbound,bean=endpoint]
    Line 18665: 2016-05-09 14:42:07.632 [main] INFO  org.springframework.integration.monitor.IntegrationMBeanExporter - Registered endpoint without MessageSource: org.springframework.integration:type=ManagedEndpoint,name=testInbound,bean=endpoint
    Line 21104: 2016-05-09 14:42:07.966 [main] DEBUG o.s.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'testInbound'
    Line 23602: 2016-05-09 14:42:09.959 [main] DEBUG o.s.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'testInbound'
    Line 23692: 2016-05-09 14:42:10.000 [main] DEBUG org.springframework.context.support.DefaultLifecycleProcessor - Starting bean 'testInbound' of type [class org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter]
    Line 23705: 2016-05-09 14:42:10.048 [main] INFO  o.s.integration.amqp.inbound.AmqpInboundChannelAdapter - started testInbound
    Line 23708: 2016-05-09 14:42:10.048 [main] DEBUG org.springframework.context.support.DefaultLifecycleProcessor - Successfully started bean 'testInbound'

**Comment by artembilan** [comment]
H-m-m. Maybe there is some service in your application which changes `autoStartup` to `true` in its `afterPropertiesSet()`?...
Would you mind to share some small application to play from our side?
I really can't reproduce it even with JMX enabled like your logs demonstrates...
I see that you use there some MVC. Maybe that is a point from somewhere to initiate `DefaultLifecycleProcessor.start()`...
Or at least, share more those DEBUG logs just exactly from the beginning of application stsatup.

**Comment by TabberGit** [comment]
Hi Artem.

I have cut down all of the non-applicable stuff from this little sample (what I was showing before was a side-piece to a larger started app)

We are working off of the spring dependency management, and use an extended version of the spring-boot-starter(s), that give us added packages such as those offered with the spring netflix oss packages and such.

It may be best if I send you a fat-jar compilation, so you can see all of the depedency libs & such.

I can also set you up with the sourcecode for the sample app itself, as well as some extensive logs with it running locally.

Where is the best place to post this info for you?

Thanks again.

**Comment by artembilan** [comment]
Hello @TabberGit !

Thank you for your patience, but I really still don't understand where is the problem...

If you can make some GitHub repository which I can fork and run locally from the IDE in the debug mode, it would be very cool!

I think GitHub even allow to restrict access by `private` repo, but I have never done that before. 
That is for case if you don't like that your code will see some one else.

**Comment by TabberGit** [comment]
Can/Will do.

I narrowed it down...

The culprit seems to be:

compile('org.springframework.cloud:spring-cloud-starter-bus-amqp:1.1.0.BUILD-SNAPSHOT')

when that is brought into the dependencies, the endpoint starts regardless of the `autoStartup(false)`.

I'll have a repo for you shortly.

**Comment by mbogoevici** [comment]
This looks like an undesirable consequence of Bus using Stream. Stream apps will start the application automatically - via https://github.com/spring-cloud/spring-cloud-stream/blob/master/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/OutputBindingLifecycle.java#L65. 

We will need to make that optional for the bus.

**Comment by TabberGit** [comment]
Artem, I've put the example code out in:

https://github.com/TabberGit/spring-integration-issue

As you'll see, I did leave logging out of the copy that I put out there.

In the build.gradle file, i have the dep. callout of the culprit.  If you write some records to a queue named `test.queue` on a default RabbitMQ config, it will start to consume them only after the /start endpoint has been hit (default server port 8080).

If you add the package in question, it will misbehave.

**Comment by TabberGit** [comment]
Thanks @mbogoevici !  Any chance something like that could be snuck in before this goes to release?

**Comment by mbogoevici** [comment]
We just released 1.0.0.RELEASE earlier, unfortunately, but if you can live with this, a 1.0.1.RELEASE for Spring Cloud Stream will follow soon. Sorry for the incovenience.

**Comment by TabberGit** [comment]
No worries!  That's fantastic.  Thanks for all of your support guys!

I think this can probably be closed if you two agree.

**Comment by mbogoevici** [comment]
Yes, I think that this belongs with the open issue in Stream. Thanks!

**Comment by artembilan** [comment]
Yeah, yeah.
Thank you @mbogoevici for such a turnaround :smile: 

Closing in favor of https://github.com/spring-cloud/spring-cloud-stream/issues/525.
The mentioning of the Spring Cloud Bus did the stuff!

Thank you @TabberGit for spotting an issue!

**Comment by SonalGuptaGemini** [comment]
Hello everyone.. 
I am facing a kind of similar issue in  my code.

I have a JmsMessageDrivenChannelAdapter which wraps up a DefaultMessageListenerContainer and sets the autostartup as false. 
eg -
	@Bean
	@Qualifier("endpointAdapter")
	public JmsMessageDrivenChannelAdapter endpointAdapter(....){
               return Jms.messageDriverChannelAdapter(endPointContainer).autoStartup(false).get();
        }

Although the DefaultLifecycleProcessor does not starts the adapter(endpointAdapter) but when I push a message on the queue binded with the container(endPointContainer), my code is still picking up the message from the queue..

Can anyone please help me as I am totally stuck here... :( :(

Thanks
